firmware/spade/src/rpi/main.c

```diff
@@ -174,9 +175,56 @@ static void rng_init(void) {
   srand(seed);
 }
 
+char serial_commands[][128] = {
+        "UPLOAD",
+        "VERSION",
+        {1, 2, 3, 4, '\0'}
+};
+
+static int read_command() {
+    int serial_command_indexes[] = {0, 0, 0};
+
+    int timeout = 0;
+
+    for (;;) {
+        int c = getchar_timeout_us(timeout);
+        if (c == PICO_ERROR_TIMEOUT) return -1;
+
+        timeout = 100;
+
+        int moved = 0;
+
+        for (int i = 0; i < ARR_LEN(serial_commands); i++) {
+            if (strlen(serial_commands[i]) > serial_command_indexes[i]
+                && serial_commands[i][serial_command_indexes[i]] == c) {
+                serial_command_indexes[i]++;
+                moved = 1;
+            }
+            if (strlen(serial_commands[i]) == serial_command_indexes[i]) {
+                return i;
+            }
+        }
+        if (!moved) return -1;
+    }
+}
+
 // Wait for a game to be uploaded.
 static int load_new_scripts(void) {
-  return upl_stdin_read();
+    int current_command = read_command();
```

Comment:
ok so `read_command` is supposed to find out what command we're getting to execute?